base/lnls-prune-artifacts.sh

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env bash
+
+set -Eeu
+
+# Filter out from the $1 list of paths any exact match of a parent directory
+# from any path in the $2 exclude list.
+#
+# Both list are treated as newline-separated strings, and must consist
+# exclusively of absolute paths.
+filter_out_paths() {
+    list="$1"
+    exclude_list="$2"
+
+    while read -r path; do
+        if [ "${path:0:1}" != "/" ]; then
+            >&2 echo "error: filter_out_paths() expects absolute paths, but got '$path'"
+            exit 1
+        fi
+
+        while [ "$path" != "/" ]; do
+            list=$(echo "$list" | grep -xv "$path")
+
+            path=$(dirname "$path")
+        done
+    done <<< "$exclude_list"
+
+    echo "$list"
+}
+
+find_elf_executables() {
+    targets=$@
+
+    # Loop on entire lines to properly handle filenames with spaces
+    while read -r file; do
+        read -r -N 4 magic < "$file"
+
+        # Output only ELF binaries
+        if [ "$magic" = $'\x7fELF' ]; then
+            echo $file
+        fi
+    done < <(find $targets -type f)
+}
+
+find_linked_libraries() {
+    executables=$(find_elf_executables $@)
+
+    # Depend on the glibc-specific behavior of supporting multiple executables
+    # to be queried at once
+    linked=$(ldd $executables 2>/dev/null | grep '=>')
+
+    # We grep out not found libraries, since they cannot be kept if we don't
+    # know where they are.
+    #
+    # Final binary may be actually runnable, since rpath of another binary may
+    # pull those not-found libraries
+    found="$(echo "$linked" | grep -v "not found")"
+
+    # Get their full path
+    libs=$(echo "$found" | cut -d' ' -f 3)
+
+    echo "$libs" | sort -u
+}
+
+get_all_epics_modules() {
+    release_defs=$(grep = ${EPICS_RELEASE_FILE} | cut -d'=' -f 2)
+
+    echo "$release_defs" | grep $EPICS_MODULES_PATH
+}
+
+get_used_epics_modules() {
+    linked_libs=$(find_linked_libraries $@)
+    all_modules=$(get_all_epics_modules)
+
+    unused_modules=$(filter_out_paths "$all_modules" "$linked_libs")
+
+    filter_out_paths "$all_modules" "$unused_modules"
```

Comment:
This logic is actually broken. If we get in `$unused_modules` something including `/opt/epics/modules/motor/modules/motorMotorSim`, for instance, the second `filter_out_paths` will also remove `/opt/epics/modules/motor`, since it is a prefix of the first one. However, we don't want such "prefixes" removal in this last step, as it should be only a pure and simple set difference.

I'll fix that.